src/main/kotlin/io/github/nomisrev/routes/articles.kt

```diff
@@ -41,6 +71,35 @@ data class Comment(
   val author: Profile
 )
 
+@Resource("/article")
+data class ArticleResource(val parent: RootResource = RootResource) {
+  @Resource("/feed") data class Feed(val parent: ArticleResource = ArticleResource())
+}
+
+fun Route.articleRoutes(
+  articleService: ArticleService,
+  jwtService: JwtService,
+) {
+
+  get<ArticleResource.Feed> {
+    jwtAuth(jwtService) { (_, userId) ->
+      either {
+          val (limit, offset) = receiveCatching<ArticleWrapper<UserFeed>>().bind().article
```

Comment:
`limit` and `offset` are being passed as query paramaters, so we don't need this here.

```suggestion
```

Instead we should validate that `limitParam` is minimum `1`, and `offsetParam` should be minimum `0`.
If we want to follow the same pattern as we're doing already we can do the following:

```kotlin
// Belongs in routes/articles.kt
@JvmInline
@Serializable
value class Offset(val offset: Int)

@JvmInline
@Serializable
value class Limit(val limit: Int)

// Following code in validations.kt
private const val MIN_FEED_SIZE = 1
private const val MIN_OFFSET = 0

data class InvalidFeedOffset(override val errors: NonEmptyList<String>) : InvalidField {
  override val field: String = "feed offset"
}

data class InvalidFeedLimit(override val errors: NonEmptyList<String>) : InvalidField {
  override val field: String = "feed limit"
}

private fun Int.minSize(size: Int): EitherNel<String, Int> =
  if (this >= size) right() else "offset too small (minimum is $size)".leftNel()

fun Int.validFeedOffset(): Either<IncorrectInput, FeedOffset> =
  minSize(MIN_OFFSET)
    .map { FeedOffset(it) }
    .mapLeft { IncorrectInput(InvalidFeedOffset(it)) }

fun Int.validFeedLimit(): Either<IncorrectInput, FeedLimit> =
  minSize(MIN_FEED_SIZE)
    .map { FeedLimit(it) }
    .mapLeft { IncorrectInput(InvalidFeedLimit(it)) }

fun ArticleResource.Feed.validate(userId: UserId): Either<IncorrectInput, GetFeed> =
  zipOrAccumulate(
    offsetParam.validFeedOffset(),
    limitParam.validFeedLimit()
  ) { offset, limit -> GetFeed(userId, limit, offset) }
    .mapLeft(::IncorrectInput)
```

Reply:
I think we need more utilities for validation like this in a library 🤔